File: 2594-Minimum Time to Repair Cars.py
class Solution:
    def repairCars(self, ranks: List[int], cars: int) -> int:
        
        left, right = 1, min(ranks)*cars*cars
        
        def isOK(time):
            res = 0
            for num in ranks:
                res += isqrt(time//num)
            return res >= cars
        
        while left + 1 < right:
            mid = left + (right - left)//2
            if isOK(mid):
                right=mid
            else:
                left = mid
                
        if isOK(left):
            return left
        else:
            return right

File: C++/2594-Minimum Time to Repair Cars.cpp
class Solution {
public:
    int Ncars;
    bool isOK(vector<int>& ranks, long long maxTime) {
        int car = 0;
        for (int r:ranks) {
            car += sqrt(maxTime/r);
            if (car >= Ncars) return true;
        }
        return false;
    }
    long long repairCars(vector<int>& ranks, int cars) {
        this->Ncars = cars;

        int mn = *min_element(begin(ranks), end(ranks));

        long long left = 1, right = (long long)mn*cars*cars;
        while (left +1 < right) {
            long long mid = left + (right-left)/2;
            if (isOK(ranks, mid)) {
                right = mid;
            } else {
                left = mid;
            }
        }

        if (isOK(ranks, left)) {
            return left;
        }
        return right;
    }
};